planning/autoware_trajectory_ranker/include/autoware/trajectory_ranker/interface/data_interface.hpp

```diff
@@ -0,0 +1,136 @@
+    scores_(metrics_num, 0.0)
+  {
+  }
+
+  void set_previous_points(const std::shared_ptr<TrajectoryPoints> & previous_points)
+  {
+    previous_points_ = previous_points;
+  }
+
+  void setup(const std::shared_ptr<TrajectoryPoints> & previous_points)
+  {
+    set_previous_points(previous_points);
+  }
+
+  void compress(const std::vector<std::vector<double>> & weight)
+  {
+    if (metrics_.empty() || weight.empty()) return;
+    for (size_t i = 0; i < metrics_.size() && i < weight.size(); i++) {
+      const auto & w = weight.at(i);
+      const auto & metric = metrics_.at(i);
+      scores_.at(i) = std::inner_product(w.begin(), w.end(), metric.begin(), 0.0);
+    }
+  }
+
+  void normalize(const double min, const double max, const size_t index, const bool flip = false)
+  {
+    if (std::abs(max - min) < std::numeric_limits<double>::epsilon()) {
+      scores_.at(index) = 1.0;
+    } else {
+      scores_.at(index) =
+        flip ? (max - scores_.at(index)) / (max - min) : (scores_.at(index) - min) / (max - min);
+    }
+  }
+
+  void weighting(const std::vector<double> & weight)
+  {
+    if (weight.empty() || scores_.empty()) {
+      total_ = 0.0;
+      return;
+    }
+    const size_t size = std::min(weight.size(), scores_.size());
+    total_ = std::inner_product(weight.begin(), weight.begin() + size, scores_.begin(), 0.0);
+  }
+
+  bool feasible() const
+  {
+    if (core_data_->points->empty()) return false;
+    constexpr double epsilon = -1e-06;
+
+    const auto condition = [&epsilon](const auto & p) {
+      return p.longitudinal_velocity_mps >= epsilon;
+    };
+    return std::all_of(core_data_->points->begin(), core_data_->points->end(), condition);
+  }
+
+  void set_metric(const size_t idx, const std::vector<double> & metric)
+  {
+    metrics_.at(idx) = metric;
+  }
+
+  auto total() const -> double { return total_; };
+
+  auto score(const size_t index) const -> double { return scores_.at(index); }
+
+  auto points() const -> std::shared_ptr<TrajectoryPoints> { return core_data_->points; }
+
+  auto previous() const -> std::shared_ptr<TrajectoryPoints> { return core_data_->previous_points; }
+
+  auto original() const -> std::shared_ptr<TrajectoryPoints> { return core_data_->original; }
+
+  auto objects() const -> std::shared_ptr<PredictedObjects> { return core_data_->objects; }
+
+  auto steering() const -> std::shared_ptr<SteeringReport> { return core_data_->steering; }
+
+  auto preferred_lanes() const -> std::shared_ptr<lanelet::ConstLanelets>
+  {
+    return core_data_->preferred_lanes;
+  }
+
+  auto header() const -> Header { return core_data_->header; }
+
+  auto uuid() const -> UUID { return core_data_->generator_id; }
+
+  auto tag() const -> std::string { return core_data_->tag; }
+
+private:
+  std::shared_ptr<CoreData> core_data_;
+
+  std::shared_ptr<TrajectoryPoints> previous_points_;
+
+  std::vector<std::vector<double>> metrics_;
+
+  std::vector<double> scores_;
+
+  double total_;
```

Comment:
Since the final output is also in float32, it seems that we don’t really need the precision of double. What do you think?
https://github.com/autowarefoundation/autoware_internal_msgs/blob/main/autoware_internal_planning_msgs/msg/ScoredCandidateTrajectory.msg#L2